adrs/20241107-spot-retries.md

```diff
@@ -0,0 +1,25 @@
+# Automatic retries on spot interruptions
+
+Job failures due to spot interruptions are a nuisance. This ADR describes a system that automatically retries the job on a new instance when a spot interruption is detected.
+
+## How it works
+
+- when we receive the `workflow_job` webhook and it's `in_progress`, extract `runner_name` from the payload and add the `runs-on-workflow-job-id` tag to the instance.
+
+```
+"runner_id": 41033,
+"runner_name": "runs-on--i-062dfa237711afbd7--FlkMSZxDTu",
+"runner_group_id": 1,
+"runner_group_name": "Default"
+```
+
+- when a spot interruption is detected, add tag on instance `runs-on-workflow-job-interrupted=true` (from agent).
+- when the workflow_job is completed and the conclusion is not success, check whether the instance has the `runs-on-workflow-job-interrupted=true` tag.
+- if yes, and `runs-on-workflow-job-id` tag is present, and retry is not set to false, then retry the workflow_job (up to specified number of retries or max 2).
+
+## How to configure
+
+- if `spot=false`, then no retries will be made (since no interruptions are expected).
+- if `retry=false` or `retry=0` in the job labels, then no retries will be made.
```

Comment:
Can I suggest `retry-interrupted` instead of the more general `retry`? It's a clearer statement of intent, IMO. `retry` alone sounds like it might do something (like retrying on failure, as some do with flaky test suites), and requires reading the docs to realize it doesn't.

Alternatively...
- `retry-if-interrupted`
- `retry-on-interrupt`

But neither of those fit in with the current label patterns, I suppose.

Reply:
After a night sleep I'm thinking of altering the behavior as follows:

- if the job is on its first run attempt (i.e. new job queued), and it fails due to a spot interruption, then auto-retry.
- if the job is on its >1 run attempt (manually retried or auto-retried), always force the use of on-demand instances.

Pros:

- ensures we don't get lots of retries when AWS has very low spot capacity, i.e. jobs will automatically switch to on-demand after a first run attempt. This avoids the need for some clever spot snoozing based on spot interruption metric.
- ensures that an auto-retried job has greater chances of success. If it fails again, it will not be due to spot interruption so there is no need to auto-retry it.

Cons:

- changes current behaviour, because a manually retried job will always use an on-demand instance, compared to using spot or on-demand based on the job labels. I think it might be a pro after all, since when you retry you want to remove uncertainty and make the job succeed.

With that, I think I can then keep the `retry` label but with only 2 values:

- `retry=when-interrupted` - the default value for spot jobs.
- `retry=false` - force-disable auto-retry in all cases.

This keeps a single label for retry behaviour, with possibly more use-cases that could be added (we could imagine RunsOn asking a separate service / lambda whether a job needs retrying).

What do you think?

Reply:
I really like this new approach. I only see pros in your message. Using spot instance is an optimization and it seems perfectly fair to switch to on-demand after a single spot eviction.

Not a very strong opinion but I would call it `autoretry` to make it more explicit when reading a job definitio

Reply:
I'm definitely in favour of a "fall back to using on-demand on spot interruption" **option**, but I think it should be opt-in. IMO, retry behaviour should match the original conditions unless the user has indicated otherwise. Failing jobs often have enough variables to try and pinpoint the failure cause, and introducing more noise into that is questionable, from that perspective. 🕵️ 

That said, if the cause is clearly spot termination, an unambiguous error message reported in the job could make this less of an issue. 🤔 

Also, I recognize that automatic fallback could be reasonably argued as consistent with the behaviour when there is not spot capacity before the job runs. I don't have a decent counterargument there, so I have to accept that my position isn't particularly robust. 😮‍💨 

But as a matter of taste at least, I prefer the model of "sensible defaults and let the user dig their own grave by overriding that when they (think they) know better". 😉